include/seqan3/search/detail/policy_search_result_builder.hpp

```diff
@@ -53,7 +53,10 @@ struct policy_search_result_builder
         std::vector<search_result_type> results(internal_hits.size());
 
         for (size_t i = 0; i < internal_hits.size(); ++i)
-            results[i] = search_result_type{0, internal_hits[i]};
+        {
```

Comment:
I suggest to remove the different make_result overloads since now the features are not mutual exclusive any more.
So the design could be simply:

```cpp
search_result_type result{};
if constexpr (search_traits_type::output_index_cursor)
   result.index_cursor = internal_hits[i];
if constexpr (search_traits_type::output_text_position)
   result.pos ....
``` 

That is just to show the basic use case. We also don't need the special constructor in the search result any more because we can anyway directly change the members due to the friend declaration. That makes everything a little easier to write and to follow and to extend.
Of course, it is not as simple as above but also not very difficult to adapt I believe.

Reply:
It is of course possible but it will result in several `constexpr if ` clauses beside the simple ones you mentioned.
For once, the need of calling locate implies storing the results in two for loops, while only returning the index needs one. Also only if locate is needed, I need to sort the results before returning them. In addition, I still need a `constexpr if` for the `search_traits_type::search_single_best_hit` which, **if locate** is needed returns only the first text position. I had this all in one function before but @marehr asked me to avoid this and I think although they are not mutual exclusive anymore, they are still different in how they are treated?

Reply:
There is a proposal from @marehr he wrote me on gitter that you might like better than the current code:
(there are still several functions but I think it nicely reduces some redundancies, thanks @marehr)

<details><summary>click me</summary>
```cpp
template <typename index_cursor_t, typename query_index_t, typename callback_t>
void make_results_impl(std::vector<index_cursor_t> internal_hits, query_index_t idx, callback_t && callback)
{
    auto maybe_locate = [](auto const & cursor)
    {
        // Ich glaube wir müssen hier nie eager suchen, da es hier keinen Vorteil bringt.
        if constexpr (search_traits_type::output_requires_locate_call)
            return cursor.lazy_locate();
        else
            return std::views::single(std::tuple{0, 0})
    }();

    for (auto const & cursor : internal_hits)
    {
        for (auto && [ref_id, ref_pos] : maybe_locate(cursor))
        {
            callback(search_result_type{std::move(idx), cursor, ref_id, ref_pos});

            if constexpr (search_traits_type::search_single_best_hit)
                return;
        }
    }
}

template <typename index_cursor_t, typename query_index_t, typename callback_t>
void make_results(std::vector<index_cursor_t> internal_hits, query_index_t idx, callback_t && callback)
{
    return make_results_impl(std::move(internal_hits), idx, std::forward<callback_t>(callback));
}

template <typename index_cursor_t, typename query_index_t, typename callback_t>
    requires search_traits_type::output_requires_locate_call &&
             (!search_traits_type::search_single_best_hit)
void make_results(std::vector<index_cursor_t> internal_hits, query_index_t idx, callback_t && callback)
{
    std::vector<search_result_type> results{};
    results.reserve(internal_hits.size()); // expect at least as many text positions as cursors, possibly more

    make_results_impl(std::move(internal_hits), idx, [&results](auto && search_result)
    {
        results.push_back(std::move(results));
    });

    // sort by reference id or by reference position if both have the same reference id.
    std::sort(results.begin(), results.end(), [] (auto const & r1, auto const & r2)
    {
        return (r1.reference_id() == r2.reference_id()) ? (r1.reference_begin_pos() < r2.reference_begin_pos())
                                                        : (r1.reference_id() < r2.reference_id());
    });
    results.erase(std::unique(results.begin(), results.end()), results.end());

    for (auto && search_result : results)
        callback(std::move(search_result));
}
```
<details>

Reply:
I wouldn't do it in this PR, but I would propose the following change:

```c++
template <typename index_cursor_t, typename query_index_t, typename callback_t>
void make_results_impl(std::vector<index_cursor_t> internal_hits, query_index_t idx, callback_t && callback)
{
    auto maybe_locate = [](auto const & cursor)
    {
        // Ich glaube wir müssen hier nie eager suchen, da es hier keinen Vorteil bringt.
        if constexpr (search_traits_type::output_requires_locate_call)
            return cursor.lazy_locate();
        else
            return std::views::single(std::tuple{0, 0})
    };

    for (auto const & cursor : internal_hits)
    {
        for (auto && [ref_id, ref_pos] : maybe_locate(cursor))
        {
            callback(search_result_type{std::move(idx), cursor, ref_id, ref_pos});

            if constexpr (search_traits_type::search_single_best_hit)
                return;
        }
    }
}

template <typename index_cursor_t, typename query_index_t, typename callback_t>
void make_results(std::vector<index_cursor_t> internal_hits, query_index_t idx, callback_t && callback)
{
    return make_results_impl(std::move(internal_hits), idx, std::forward<callback_t>(callback));
}

template <typename index_cursor_t, typename query_index_t, typename callback_t>
    requires search_traits_type::output_requires_locate_call &&
             (!search_traits_type::search_single_best_hit)
void make_results(std::vector<index_cursor_t> internal_hits, query_index_t idx, callback_t && callback)
{
    std::vector<search_result_type> results{};
    results.reserve(internal_hits.size()); // expect at least as many text positions as cursors, possibly more

    make_results_impl(std::move(internal_hits), idx, [&results](auto && search_result)
    {
        results.push_back(std::move(search_result));
    });

    // sort by reference id or by reference position if both have the same reference id.
    std::sort(results.begin(), results.end(), [] (auto const & r1, auto const & r2)
    {
        return (r1.reference_id() == r2.reference_id()) ? (r1.reference_begin_pos() < r2.reference_begin_pos())
                                                        : (r1.reference_id() < r2.reference_id());
    });
    results.erase(std::unique(results.begin(), results.end()), results.end());

    for (auto && search_result : results)
        callback(std::move(search_result));
}
```

Reply:
Well, yes something like this. This is basically what I said:

> Of course, it is not as simple as above but also not very difficult to adapt I believe.

But instead of having this weird constructor in the search_result do at the respective places something like: 
```cpp
if constexpr (needs_id)
     result.id = idx;
if constexpr (needs_ref_pos)
    result.ref_pos = pos;
// ... etc.
```

Reply:
And please make it part of this PR but as a new commit.